registry/api/v2/urls.go

```diff
@@ -42,32 +42,46 @@ func NewURLBuilderFromString(root string) (*URLBuilder, error) {
 }
 
 // NewURLBuilderFromRequest uses information from an *http.Request to
-// construct the root url.
-func NewURLBuilderFromRequest(r *http.Request) *URLBuilder {
-	var scheme string
-
-	forwardedProto := r.Header.Get("X-Forwarded-Proto")
-
-	switch {
-	case len(forwardedProto) > 0:
-		scheme = forwardedProto
-	case r.TLS != nil:
-		scheme = "https"
-	case len(r.URL.Scheme) > 0:
-		scheme = r.URL.Scheme
-	default:
-		scheme = "http"
-	}
+// construct the root url. configScheme and configHost are the URI scheme and
+// hostname, respectively, from the configuration - or empty strings if
+// unspecified.
+func NewURLBuilderFromRequest(r *http.Request, configScheme, configHost string) *URLBuilder {
```

Comment:
I gave this a try and put the change in a separate PR for comparison: https://github.com/docker/distribution/pull/1035

To be honest, I like the approach in this PR more. Both use cases need the `Request` to extract the path, and the unit tests are cleaner when there's a single function that encapsulates the logic for building a URL from a request and an optional preconfigured scheme/hostname. But I'll leave it up to you - feel free to merge whichever PR you think is better (or of course, give additional feedback).